keybox/keybox.c

```diff
@@ -0,0 +1,252 @@
+
+    return 0;
+}
+
+static ssize_t read_with_retry(int fd, void *buf_, size_t size, off_t offset)
+{
+    ssize_t rc;
+    size_t  rcnt = 0;
+    uint8_t *buf = buf_;
+
+    while (size > 0) {
+        rc = TEMP_FAILURE_RETRY(pread(fd, buf, size, offset));
+        if (rc < 0)
+            return rc;
+        if (rc == 0)
+            break;
+        size -= rc;
+        buf += rc;
+        offset += rc;
+        rcnt += rc;
+    }
+    return rcnt;
+}
+
+static int read_keybox_from_teedata(const char *keybox_path, char *out, size_t *out_size)
+{
+    int rc;
+    struct keybox_header header;
+
+    rc = open(keybox_path, O_RDWR, 0);
+    if (rc < 0) {
+        ALOGE("%s: unable (%d) to open keybox file '%s': %s\n",
+              __func__, errno, keybox_path, strerror(errno));
+        return rc;
+    }
+    tee_fd = rc;
+
+    rc = read_with_retry(tee_fd, &header, sizeof(struct keybox_header), KB_HEAD_OFFSET);
+    if (rc < 0) {
+        ALOGE("%s read failed.\n", __func__);
+        return -1;
+    }
+
+    if (memcmp("MAGICKEYBOX", header.magic, sizeof(header.magic))) {
+        ALOGE("Keybox not found. exit...\n");
+        return -1;
+    }
+
+    ALOGI("keybox size is %u.\n", header.size);
+
+    if (header.size > MAX_FILE_SIZE) {
+        ALOGE("keybox size is too big %u.\n", header.size);
+        return -1;
+    }
+
+    *out_size = header.size;
+    read_with_retry(tee_fd, out, header.size, sizeof(struct keybox_header));
+
+    /* Append the end string to the keybox ending as the flag of finishing transfer */
+    memcpy(((uint8_t *)out + *out_size), end_str, sizeof(end_str));
+    *out_size += sizeof(end_str);
+
+    return 0;
+}
+
+bool provision_keybox(const char* trusty_devname, const char *keybox_path)
+{
+    int rc = -1;
+    bool ret = false;
+    keymaster_message_t *msg = NULL;
+    size_t keybox_req_size = 0;
+    size_t msg_size = 0;
+    size_t file_size = 0;
+    char buf[MAX_FILE_SIZE + sizeof(end_str)] = {0};
+    char recv_buf[PAGE_SIZE] = {0};
+    keybox_provision_req_t *keybox_req;
+    int tipc_fd = -1;
+
+    if (keybox_path == NULL)
+        return false;
+
+    rc = read_keybox_from_teedata(keybox_path, buf, &file_size);
+    if (rc < 0) {
+        ALOGE("Retrieve keybox from teedata: FAIL\n");
+        goto clean;
+    }
+
+    if (trusty_devname == NULL)
+        goto clean;
+
+    if (file_size + sizeof(keybox_provision_req_t) + sizeof(keymaster_message_t) > 16383) {
+        ALOGE("Keybox size(%zu) is over-sized.\n", file_size);
+        goto clean;
+    }
+
+    size_t max_transfer_size = PAGE_SIZE - sizeof(struct tipc_msg_hdr) - sizeof(keybox_provision_req_t) - sizeof(keymaster_message_t);
+    size_t remaining_size = file_size;
+    size_t transfer_size = 0;
+    uint8_t* pbuf = (uint8_t*)buf;
+
+    ALOGI("keybox max_transfer_size is %zu.\n", max_transfer_size);
+
+    while (remaining_size)
+    {
+        /* Connect to Keymaster TA */
+        tipc_fd = tipc_connect(trusty_devname, KEYMASTER_PORT);
```

Comment:
better to move the connect/close out of the while().

Reply:
let it go.  otherwise, it will fail during the 2nd send.